src/test/java/org/broadinstitute/hellbender/utils/read/ReadUtilsUnitTest.java

```diff
@@ -387,4 +390,47 @@ public void testGetReferenceIndex( final GATKRead read, final SAMFileHeader head
         Assert.assertEquals(ReadUtils.getReferenceIndex(read, header), expectedReferenceIndex, "Wrong reference index for read");
         Assert.assertEquals(ReadUtils.getMateReferenceIndex(read, header), expectedMateReferenceIndex, "Wrong reference index for read's mate");
     }
+
+    @DataProvider(name="createSAMWriter")
+    public Object[][] createSAMWriterData() {
+        return new Object[][] {
+                // Note: count_bases==query sorted, unindexed==coord sorted. We expect to silently fail
+                // to create an index if createIndex is true and sort order is not coord.
+                {new File(getToolTestDataDir(), "count_bases.bam"),   false,  true, true, false},
+                {new File(getToolTestDataDir(), "unindexed.bam"),     false,  true, true, true},
+                {new File(getToolTestDataDir(), "count_bases.bam"),   true,   true, true, false},
+                {new File(getToolTestDataDir(), "unindexed.bam"),     true,   true, true, true},
+                {new File(getToolTestDataDir(), "count_bases.bam"),   true,   true, false, false},
+                {new File(getToolTestDataDir(), "unindexed.bam"),     true,   true, false, true},
+        };
+    }
+
+    @Test(dataProvider="createSAMWriter")
+    public void testCreateSAMWriter(
+            final File bamFile,
+            final boolean preSorted,
+            final boolean createIndex,
+            final boolean createMD5,
+            final boolean expectIndex) throws Exception {
+
+        final File outputFile = createTempFile("samWriterTest",  ".bam");
+
+        try (final SamReader samReader = SamReaderFactory.makeDefault().open(bamFile)) {
+             final SAMFileHeader header = samReader.getFileHeader();
+            // ensure test condition
+            Assert.assertEquals(expectIndex, header.getSortOrder() == SAMFileHeader.SortOrder.coordinate);
+
+            try (final SAMFileWriter samWriter = ReadUtils.createCommonSAMWriter
+                            (outputFile, null, samReader.getFileHeader(), preSorted, createIndex, createMD5)) {
+                final Iterator<SAMRecord> samRecIt = samReader.iterator();
+                while (samRecIt.hasNext()) {
+                    samWriter.addAlignment(samRecIt.next());
+                }
+            }
+        }
+
+        Assert.assertEquals(expectIndex, null != SamFiles.findIndex(outputFile));
+        final File md5File = new File(outputFile.getAbsolutePath() + ".md5");
+        Assert.assertEquals(createMD5, md5File.exists());
```

Comment:
Set the `.md5` file to be deleted on exit (note that `createTempFile()` already marks the bam index for deletion on exit for you, but it knows nothing about these `.md5` files).